src/i_oalmusic.c

```diff
@@ -196,24 +194,20 @@ static int PlayerThread(void *unused)
 {
     SDL_SetThreadPriority(SDL_THREAD_PRIORITY_TIME_CRITICAL);
 
-    SDL_LockMutex(music_lock);
     StartPlayer();
-    SDL_UnlockMutex(music_lock);
 
     while (player_thread_running)
     {
         boolean result;
 
-        SDL_LockMutex(music_lock);
         result = UpdatePlayer();
-        SDL_UnlockMutex(music_lock);
 
         if (result == false)
         {
             break;
         }
 
-        SDL_Delay(100);
+        SDL_Delay(10);
```

Comment:
> With this change, there is still a brief pause (game freezes for a moment) but it's not as severe as before.

Now the only change from Woof 12.0.2 is an increased delay and some clean up in `I_OAL_StopSong`. Try `SDL_Delay(1)`.

Reply:
> Try `SDL_Delay(1)`.

That does reduce the brief pause even further.

Reply:
So should we reduce it to `SDL_Delay(1)`? To be honest, I can't see any of these delays on my 1080p60 laptop 😄

Reply:
I don't know, but whatever solution is used needs to eliminate the hitching completely. I find it to be a big issue, personally. The opening room with the cyberdemon in aaliens becomes unplayable with even the slightest pause, so we shouldn't settle for anything less than perfect. I wonder how other source ports handle that initial music change.

Reply:
I reduce it here: https://github.com/fabiangreffrath/woof/pull/1430/commits/39ff73a5f8a5f30d046b457d2918cc897b57eade

Reply:
> I wonder how other source ports handle that initial music change and if they have a slight pause as well.

GZDoom is terrible, Eternity switched to i_winmuisc.c

Reply:
> > I wonder how other source ports handle that initial music change and if they have a slight pause as well.
> 
> GZDoom is terrible, Eternity switched to i_winmuisc.c (SDL_Mixer delay is terrible)

Reply:
Just some ideas to help brainstorm: caching all the songs for a level in advance, initializing both the music file (ogg, etc.) and midi systems and keeping them idle and ready even if one or the other is not being used currently, moving more processing to a separate thread.

Reply:
> GZDoom is terrible, Eternity switched to i_winmusic.c (SDL_Mixer delay was very bad too)

I just tried DSDA and it doesn't have any delay with portmidi or fluidsynth.

Reply:
> initializing both the music file (ogg, etc.) and midi systems and keeping them idle and ready even if one or the other is not being used currently

This is what we are already doing.

> moving more processing to a separate thread.

We can move MIDI/OGG etc loading into the player thread.

Is this PR now at Woof 12.0.2 level?

Reply:
> I just tried DSDA and it doesn't have any delay with portmidi or fluidsynth.

On song change we wait for oal stream thread or winmusic thread, DSDA has different thread system. I suggest to improve it later, for me the current delay is OK (I can't see it).

Reply:
> Is this PR now at Woof 12.0.2 level?

Yes, they seems to be the same now. There is a very small pause that probably depends on how much of the 1 ms delay has elapsed so it's almost imperceptible. If no one has complained about it and we're back at the 12.0.2 level it appears to be a non-issue.

Reply:
I think these are good topics to investigate in the future, but we're okay for now.

> caching all the songs for a level in advance

> We can move MIDI/OGG etc loading into the player thread.

> Currently, OAL Stream processes many samples in advance.

Reply:
> There is a very small pause that probably depends on how much of the 1 ms delay has elapsed so it's almost imperceptible. 


Try setting `NUM_BUFFERS 2` and `BUFFER_SAMPLES 1024` to further reduce the delay.
 https://github.com/fabiangreffrath/woof/blob/4d3d5b0fef69c6ff8dae61f3374b892e93b50380/src/i_oalmusic.c#L33-L34

Reply:
> Try setting NUM_BUFFERS 2 and BUFFER_SAMPLES 1024 to further reduce the delay.

I tried it and it doesn't seem to make any difference. I'm guessing that has more to do with reducing the audio latency slightly.